sycl/source/detail/platform_impl.cpp

```diff
@@ -141,14 +142,26 @@ std::vector<platform> platform_impl::get_platforms() {
 // This function matches devices in the order of backend, device_type, and
 // device_num.
 static void filterDeviceFilter(std::vector<RT::PiDevice> &PiDevices,
```

Comment:
It seems there are still races between calls to new `plugin` functions. See next suggestions.

Reply:
Thanks. I committed all your suggested code changes.

Reply:
@romanovvlad ,
Unfortunately, your code changes caused lots of tests to fail.
The reason is that the global locks you put in `get_platforms` and `get_devices` to be deadlocked and DeviceId and PlatformId are not thread-safe.
I clarified the purpose of the global lock and plugin-internal locks in the comment to help you understand how we should use them.

- Global `MPluginsMutex` is intended to guard only the `MPlugins` vector itself.

- Local `MDeviceIDMutex` and `MPlatformIdMutex` are intended to guard internal data members in the `plugin` class.

So, I reverted all your changes to make tests to pass.
Let me know if you want to change it differently.